### ISSUE TITLE
Build Error

### ISSUE THREAD

**Issue by Ryvian** [Nashenas88/borrow_visualizer_prototype#1]
Hi Nashenas88!

~~I am trying to use this prototype, but a problem occurs when building:~~
![image](https://user-images.githubusercontent.com/31315697/61166340-a72bfa80-a4f9-11e9-8143-d5e4d1fcc555.png)

~~That is, even though I do these steps successfully:~~
![image](https://user-images.githubusercontent.com/31315697/61166357-ed815980-a4f9-11e9-8bed-0fe2595df31b.png)

~~I still meet this error when "cargo build". Can you figure out what's wrong here?~~

Thanks.

--------
well problem solved. made a stupid mistake here

**Comment by Ryvian** [comment]
By the way, I am now working on a project on Rust, and could you tell me whether this tool will give a valid lifetime display on IDE (such as Atom or VSCode) for a temporary lock object? In the following example, this is  "client.read()" object:
```rust
fn do_request() {
    //client: Arc<RwLock<Inner>>
    match connect(client.read().unwrap().m) {
        Ok(_) => {
            let mut inner = client.write().unwrap();
            inner.m = mbrs;
        }
        Err(_) => {}
    }
}
```

The "client" object is an "inner" object wrapped by RwLock, and what I want the IDE to display is the lifetime of "client.read()". As you see, "client.read()" 's lifetime spans from line 3 to the end of match. Is this tool able to display this?

As I am unable to install this tool so far I can't validate it by myself. Many thanks if you can answer this question!

**Comment by Ryvian** [comment]
@Nashenas88

**Comment by Ryvian** [comment]
Well, I still cannot compile it:
```
error[E0432]: unresolved imports `rustc_driver::driver`, `rustc_driver::CompilerCalls`
  --> src\borrowcalls.rs:21:20
   |
21 | use rustc_driver::{driver, CompilerCalls, Compilation};
   |                    ^^^^^^  ^^^^^^^^^^^^^ no `CompilerCalls` in the root
   |                    |
   |                    no `driver` in the root

error[E0433]: failed to resolve: could not find `Expr_` in `hir`
   --> src\borrowcalls.rs:295:14
    |
295 |         hir::Expr_::ExprAddrOf(_, ref e) => expr_borrows_node_id(tcx, id, &*e, true),
    |              ^^^^^ could not find `Expr_` in `hir`

error[E0433]: failed to resolve: could not find `Expr_` in `hir`
   --> src\borrowcalls.rs:296:14
    |
296 |         hir::Expr_::ExprMatch(ref e, ref arms, ..) => {
    |              ^^^^^ could not find `Expr_` in `hir`

error[E0433]: failed to resolve: could not find `Expr_` in `hir`
   --> src\borrowcalls.rs:304:14
    |
304 |         hir::Expr_::ExprTup(ref vec) => vec.iter().any(|e| expr_borrows_node_id(tcx, id, e, false)),
    |              ^^^^^ could not find `Expr_` in `hir`

error[E0433]: failed to resolve: could not find `Expr_` in `hir`
   --> src\borrowcalls.rs:305:14
    |
305 |         hir::Expr_::ExprPath(hir::QPath::Resolved(_, ref path)) if is_ref => {
    |              ^^^^^ could not find `Expr_` in `hir`

error[E0433]: failed to resolve: could not find `Def` in `def`
   --> src\borrowcalls.rs:308:27
    |
308 |                 hir::def::Def::Local(def_id) |
    |                           ^^^ could not find `Def` in `def`

error[E0433]: failed to resolve: could not find `Def` in `def`
   --> src\borrowcalls.rs:309:27
    |
309 |                 hir::def::Def::Upvar(def_id, ..) => {
    |                           ^^^ could not find `Def` in `def`

error[E0433]: failed to resolve: could not find `codemap` in `syntax`
   --> src\borrowcalls.rs:329:24
    |
329 |     let span = syntax::codemap::original_sp(input_span, wrapping_span);
    |                        ^^^^^^^ could not find `codemap` in `syntax`

error[E0531]: cannot find tuple struct/variant `NodeExpr` in module `hir_map`
   --> src\borrowcalls.rs:204:41
    |
204 |                         if let hir_map::NodeExpr(expr) = borrow_node {
    |                                         ^^^^^^^^ not found in `hir_map`

error[E0531]: cannot find tuple struct/variant `NodeStmt` in module `hir_map`
   --> src\borrowcalls.rs:369:29
    |
369 |             if let hir_map::NodeStmt(..) = node {
    |                             ^^^^^^^^ not found in `hir_map`

error[E0531]: cannot find tuple struct/variant `NodeBlock` in module `hir_map`
   --> src\borrowcalls.rs:388:26
    |
388 |                 hir_map::NodeBlock(_) => continue,
    |                          ^^^^^^^^^ not found in `hir_map`

error[E0531]: cannot find tuple struct/variant `NodeLocal` in module `hir_map`
   --> src\borrowcalls.rs:389:26
    |
389 |                 hir_map::NodeLocal(_) => {
    |                          ^^^^^^^^^ not found in `hir_map`

error[E0531]: cannot find tuple struct/variant `NodePat` in module `hir_map`
   --> src\borrowcalls.rs:395:26
    |
395 |                 hir_map::NodePat(ref pat) => {
    |                          ^^^^^^^ not found in `hir_map`

error[E0603]: enum `Node` is private
   --> src\borrowcalls.rs:338:57
    |
338 |         -> Option<(syntax::ast::NodeId, Option<hir_map::Node<'tcx>>, blocks::FnLikeNode<'tcx>, hir_map::Node<'tcx>, syntax::ast::NodeId)> {
    |                                                         ^^^^

error[E0603]: enum `Node` is private
   --> src\borrowcalls.rs:338:105
    |
338 |         -> Option<(syntax::ast::NodeId, Option<hir_map::Node<'tcx>>, blocks::FnLikeNode<'tcx>, hir_map::Node<'tcx>, syntax::ast::NodeId)> {
    |                                                                                                         ^^^^

error[E0603]: enum `Node` is private
   --> src\borrowcalls.rs:423:102
    |
423 | fn get_fn_node<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, id: syntax::ast::NodeId, node: Option<hir_map::Node<'tcx>>)
    |                                                                                                      ^^^^

error[E0603]: enum `Node` is private
   --> src\borrowcalls.rs:424:57
    |
424 |         -> Option<(syntax::ast::NodeId, Option<hir_map::Node<'tcx>>, blocks::FnLikeNode<'tcx>, hir_map::Node<'tcx>, syntax::ast::NodeId)> {
    |                                                         ^^^^

error[E0603]: enum `Node` is private
   --> src\borrowcalls.rs:424:105
    |
424 |         -> Option<(syntax::ast::NodeId, Option<hir_map::Node<'tcx>>, blocks::FnLikeNode<'tcx>, hir_map::Node<'tcx>, syntax::ast::NodeId)> {
    |                                                                                                         ^^^^

warning: unused import: `rustc::dep_graph::DepNode`
  --> src\borrowcalls.rs:15:5
   |
15 | use rustc::dep_graph::DepNode;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: #[warn(unused_imports)] on by default

error: aborting due to 18 previous errors

Some errors have detailed explanations: E0432, E0433, E0603.
For more information about an error, try `rustc --explain E0432`.
error: Could not compile `borrow_bounds`.

To learn more, run the command again with --verbose.
```

Does anyone know what happen here? Thanks.

**Comment by Nashenas88** [comment]
@lyrii so sorry for the late reply! This was buried amongst a lot of other notifications.

This project hasn't been supported for over 2 years now. It was heavily dependent on unsupported compiler internals that have changed drastically in that time period. Given the additional fact that this project doesn't even take NLL into consideration (it was in the very early stages of development when I stopped working on this), it would essentially require a rewrite to get this project working again. If you're interested in the latest updates (I'm only just about getting ready to start back into the rust community again), follow along this thread: https://internals.rust-lang.org/t/borrow-visualizer-for-the-rust-language-service/4187

**Comment by Ryvian** [comment]
OK. I am curious how did you get the lifetime information? Is there any convenient rustc API or other things to do that? Can RLS give this info?
@Nashenas88

**Comment by Nashenas88** [comment]
Unfortunately there was no convenient API available. I had maintained a fork of the compiler that exposed some internals publicly, but that no longer functions either. A big reason I stopped working on this specific project was because of how much time I would have to spend fixing merge conflicts and dealing with redesigns in the compiler. More than half my time working on this project was spent dealing with merge issues and API changes.

Mentioned near the bottom of the internals thread I posted above, was a plan to come up with ideas for the best visualizations. Then work backwards to see what would need to be exposed from the compiler to support such a change.

If you'd like to move forward with that path, I would recommend following that internals thread or starting a new one.
